[PF-278] First pass at Kubernetes common library

tnadler opened the pull request.

tnadler committed:
First pass at Kubernetes common library

tnadler committed:
Merge branch 'develop' into tn-PF-278-kubernetes-lib

ddietterich committed:
Remove unused variable; add documentation

ddietterich committed:
Make KubeService a non-@Component class; remove unneeded classes

ddietterich committed:
Merge from develop

ddietterich committed:
Remove unused methods, use concurrent map, clean up variable names

jaycarlton commented on build.gradle:
```diff
@@ -23,7 +23,7 @@ repositories {
 group 'bio.terra'
 version '0.0.15-SNAPSHOT'
-sourceCompatibility = JavaVersion.VERSION_11
+sourceCompatibility = JavaVersion.VERSION_1_8
```
Maybe this change by itself should be its own PR, round of testing, and version change.

jaycarlton commented on src/main/java/bio/terra/common/kubernetes/KubePodListener.java:
```diff
@@ -0,0 +1,184 @@
+  private final String podNameFilter;
+  private final Stairway stairway;
+  private Exception exception;
+  private final Map<String, Boolean> podMap; // pod name; true means running; false means deleted
```
Please use a `PodState` enum instead of a boolean for readability and expandability.

yonghaoy commented on src/main/java/bio/terra/common/kubernetes/KubeService.java:
```diff
@@ -0,0 +1,195 @@
+   *     name, so components must plumb this in using settings in the helm charts.
+   * @param inKubernetes Used to denote that we are running in the Kubernetes environment. The
+   *     service can be called and provides reasonable answers when inKubernetes is false.
+   * @param podNameFilter Filter to apply to pod names returned by Kubernetes API to know which pods
+   *     we care about.
+   */
+  public KubeService(String podName, boolean inKubernetes, String podNameFilter) {
```
I am not 100% sure thow to use podNameFilter and podName. 
Draft a PR: https://github.com/broadinstitute/terra-helm/pull/286/files

```
podName: metadata.name
podNameFilter: the one in line 25?
```
Can you help to take look to see if my understanding is correctly. 
Thanks

ddietterich replied:
Your setup of `podName` and `inKubernetes` look right to me. That is what TDR does [here](https://github.com/broadinstitute/datarepo-helm/blob/master/charts/datarepo-api/templates/api-deployment.yaml)

In TDR, the podNameFilter is a constant `datarepo-api` because that substring is used  for the main DRmanager code in all deployments. So long as all Buffer Service pods in a deployment have `{{ .Values.name }}-service` as part of their name (and no other pod has it as part of its name), then it should work.

yonghaoy commented on src/main/java/bio/terra/common/kubernetes/KubeService.java:
```diff
@@ -0,0 +1,195 @@
+      for (V1Pod item : list.getItems()) {
+        if (item.getMetadata() != null) {
+          String podName = item.getMetadata().getName();
+          if (StringUtils.contains(podName, podNameFilter)) {
```
In the original TDR code, you have `getPods()` `getApiPods()`, the 2nd one uses `podNameFilter`. 
Do we need to keep the original `getPods()`, that does not have the filter?

ddietterich replied:
I am not finding the `getPods()` you are talking about, so not understanding the question. I do not have a use case for exposing the whole pod list.

Now that I look at it, I have not done a good job making things package-private. I think I should do one more interface revision.

yonghaoy commented on src/main/java/bio/terra/common/kubernetes/KubeService.java:
```diff
@@ -0,0 +1,195 @@
+    }
+    int defaultPodCount = 1;
+    logger.info("KubeService ActivePodCount - default val: {}", defaultPodCount);
+    return defaultPodCount;
```
Does this always return 1?

ddietterich replied:
Only if we are not running under Kubernetes

ddietterich replied:
No. It returns 1 if `inKubernetes` is false.

yonghaoy commented on src/main/java/bio/terra/common/kubernetes/KubeService.java:
```diff
@@ -0,0 +1,195 @@
+      return new String(encoded, StandardCharsets.UTF_8);
+    } catch (IOException e) {
+      logger.error("Failed to read file: " + path + "; ", e);
+      return null;
```
Client may not be aware `getNamespace()` may return null. Feel it is possible to cause NPE in client code. Suggest to use `Optional`, or throw exception there?

yonghaoy replied:
Did you push you commit to address this?

ddietterich replied:
Not yet. I should have this updated by days end.

yonghaoy reviewed (commented):
Thanks! Few questions!

jaycarlton commented on src/main/java/bio/terra/common/kubernetes/KubePodListener.java:
```diff
@@ -0,0 +1,184 @@
+      try {
+        ApiClient client = Config.defaultClient();
+        // infinite timeout
+        OkHttpClient httpClient =
```
The names here are confusing me. Maybe inline `httpClient`.

ddietterich committed:
Make pod listener package private
Consolidate shutdown state and stopPodListener
Throw error if file cannot be read

jaycarlton commented on src/main/java/bio/terra/common/kubernetes/KubePodListener.java:
```diff
@@ -0,0 +1,184 @@
+        try (Watch<V1Namespace> watch =
+            Watch.createWatch(
+                client,
+                api.listNamespacedPodCall(
```
Is there a way to make an argument object that has a builder? This is a lot of nullable arguments.

ddietterich replied:
Yes, the Kubernetes Java client API is crazy with nulls.

jaycarlton commented on src/main/java/bio/terra/common/kubernetes/KubePodListener.java:
```diff
@@ -0,0 +1,184 @@
+                    podMap.put(podName, false);
+                  }
+                } catch (DatabaseOperationException | StairwayExecutionException ex) {
+                  logger.error("Stairway recoveryStairway failed to recovery pod: " + podName, ex);
```
*recover

jaycarlton commented on src/main/java/bio/terra/common/kubernetes/KubePodListener.java:
```diff
@@ -0,0 +1,184 @@
+    return exception;
+  }
+
+  Map<String, Boolean> getPodMap() {
```
If you don't want to make this enum-valued, then just a set of running pods would work I would think.

jaycarlton commented on src/main/java/bio/terra/common/kubernetes/KubeService.java:
```diff
@@ -0,0 +1,190 @@
+  // Location in the container where Kubernetes stores service account and
+  // namespace information. Kubernetes mints a service account for the container (pod?)
+  // that can be used to make requests of Kubernetes.
+  static final String KUBE_DIR = "/var/run/secrets/kubernetes.io/serviceaccount";
```
should this be `KUBE_SA_DIR`?

jaycarlton commented on src/main/java/bio/terra/common/kubernetes/KubeService.java:
```diff
@@ -0,0 +1,190 @@
+   *
+   * @return set of pod names containing the podNameFilter string; null if not in kubernetes
+   */
+  public Set<String> getPodList() {
```
name this `getPodSet()` or `getPods()`

jaycarlton commented on src/main/java/bio/terra/common/kubernetes/KubeService.java:
```diff
@@ -0,0 +1,190 @@
+    try {
+      CoreV1Api api = makeCoreApi();
+      V1PodList list =
+          api.listNamespacedPod(namespace, null, null, null, null, null, null, null, null, null);
```
Do we control this api? Can we overload `listNamespacePod`?

ddietterich replied:
No, we do not control the api.

jaycarlton commented on src/main/java/bio/terra/common/kubernetes/KubeService.java:
```diff
@@ -0,0 +1,190 @@
+  }
+
+  public int getActivePodCount() {
+    if (podListener != null) {
```
I'd use `Optional.ofNullable().map().orElse()` here.

ddietterich replied:
I would never use that. Why is everyone so twitchy about comparing to null? What is the benefit of running through 4 methods to do the same thing?

jaycarlton replied:
It's more deliberate, and a higher level of abstraction. There's no need to branch, and you're very clear about what you're getting and that it might not be there and what the default is.

`if` is basically deprecated as of Java 8.

jaycarlton commented on src/main/java/bio/terra/common/kubernetes/KubeService.java:
```diff
@@ -0,0 +1,190 @@
+   * @param joinWait time in the time unit to wait
+   * @return true if we joined the listener thread; false if that timed out
+   */
+  public boolean stopPodListener(TimeUnit timeUnit, long joinWait) {
```
`Duration` is easier to work with.

ddietterich replied:
This expression is a common idiom.

jaycarlton commented on src/main/java/bio/terra/common/kubernetes/KubeService.java:
```diff
@@ -0,0 +1,190 @@
+
+  private CoreV1Api makeCoreApi() {
+    try {
+      ApiClient client = ClientBuilder.cluster().build();
```
This is confusing me. You make one kind of client and then return a different one?

ddietterich replied:
This is how the K8s API manages its different features. You build different sub-clients for each thing you want to do. So there are alpha clients and beta clients. All kinds of clients.

jaycarlton reviewed (approved):
is there a way to write unit tests for this code?

ddietterich commented:
@jaycarlton I do not know a useful way to unit test this. Its entire purpose is interacting with Kubernetes.
I have been testing it by building data repo with it and running the repo integration tests that run under Kubernetes.

ddietterich committed:
Updates from review
